src/lib/query/pci.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{Iface, NisporError};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(into = "String")]
+#[serde(try_from = "String")]
+pub struct PciAddress {
+    pub domain: u32,
+    pub bus: u8,
+    pub slot: u8,
+    pub function: u8,
+}
+
+impl std::fmt::Display for PciAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}:{:02x}:{:02x}.{}",
+            if self.domain > u16::MAX.into() {
+                format!("{:08x}", self.domain)
+            } else {
+                format!("{:04x}", self.domain)
+            },
+            self.bus,
+            self.slot,
+            self.function
+        )
+    }
+}
+
+impl TryFrom<String> for PciAddress {
+    type Error = NisporError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::from_str(value.as_str())
+    }
+}
+
+impl From<PciAddress> for String {
+    fn from(v: PciAddress) -> String {
+        v.to_string()
+    }
+}
+
+impl FromStr for PciAddress {
+    type Err = NisporError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut items: Vec<&str> = s.split(":").collect();
+        let err = NisporError::invalid_argument(format!(
```

Comment:
For USB interfaces the PCI address is something like `2-2.4.4.3:1.0`. I don't see this error for a USB or netdevsim interface (which uses `netdevsim1` as PCI address).

Reply:
I do not have plan to support USB interface.